yarn-project/noir-contracts/src/contracts/lending_contract/src/main.nr

```diff
@@ -13,137 +23,135 @@ contract Lending {
-        // Return private circuit public inputs. All private functions need to return this as it is part of the input of the private kernel.
+    fn constructor(inputs: PrivateContextInputs) -> distinct pub abi::PrivateCircuitPublicInputs {
         PrivateContext::new(inputs, 0).finish()
     }
 
     open fn init(
-        inputs: PublicContextInputs
+        inputs: PublicContextInputs,
+        oracle_address: Field,
+        loan_to_value: Field,
+        collateral_asset: Field,
+        stable_coin: Field,
     ) -> pub abi::PublicCircuitPublicInputs {
-        let mut context = PublicContext::new(inputs, abi::hash_args([]));
-
+        let mut context = PublicContext::new(inputs, abi::hash_args([oracle_address, loan_to_value, collateral_asset, stable_coin]));
         let storage = Storage::init();
+        let asset_loc = storage.assets.at(0);
+        let asset = asset_loc.read();
 
-        let asset = storage.assets.at(0);
+        assert (loan_to_value as u120 <= 10000);
+        assert (asset.last_updated_ts == 0);
+        assert (asset.interest_accumulator == 0);
 
-        let tot = asset.read();
-
-        assert (tot.last_updated_ts == 0);
-        assert (tot.interest_accumulator == 0);
-
-        asset.write(Tot {
+        asset_loc.write(Asset {
             interest_accumulator: 1000000000,
             last_updated_ts: context.timestamp() as u120,
+            loan_to_value: loan_to_value as u120,
+            oracle_address,
         });
 
+        storage.collateral_asset.write(collateral_asset);
+        storage.stable_coin.write(stable_coin);
+
         context.return_values.push(1);
         context.finish()
     }
-
-    // Create a position.
-
-    open fn update_tot(
-        inputs: PublicContextInputs
-    ) -> pub abi::PublicCircuitPublicInputs {
+    // keccak256("update_accumulator()") >> 224 -> 0x1873b536
+    open fn update_accumulator(inputs: PublicContextInputs) -> pub abi::PublicCircuitPublicInputs {
         let mut context = PublicContext::new(inputs, abi::hash_args([]));
         let storage = Storage::init();
 
-        let asset = storage.assets.at(0);
-        let mut tot = asset.read();
+        let asset_loc = storage.assets.at(0);
+        let mut asset = asset_loc.read();
 
-        let dt: u120 = context.timestamp() as u120  - tot.last_updated_ts;
+        let dt: SafeU120 = SafeU120{value: context.timestamp() as u120}.sub(SafeU120{value: asset.last_updated_ts});
 
-        // Need proper interest computation here.
-        // Assume constant rate because too few reads :cry:
-        let rate: u120 = 100000;
-        // Just need something that increases, but :cry:
-        let multiplier: u120 = 1000000000 + dt * rate;
-        let divisor: u120 = 1000000000;
+        // Only update if time has passed.
+        if (!dt.is_zero()) {
+            let precision: SafeU120 = SafeU120{value: 1000000000};
+            let rate_per_second: u120 = 1268391679; // 4% yearly rate / (60 * 60 * 24 * 365)
```

Comment:
Practically yes as it is done now. Ideally it would be part of the `Asset` but then the `update_accumulator` cannot return the entire asset as it is larger than `RETURN_VALUES_LENGTH = 4` as long as I am not doing storage packing.